pkg/cli/ignition/convert3.go

```diff
@@ -0,0 +1,116 @@
+package ignition
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/coreos/ign-converter/translate/v24tov31"
+	ignv2 "github.com/coreos/ignition/config/v2_4"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	kcmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/util/templates"
+)
+
+const convert3CommandName = "convert3"
+
+type convert3Options struct {
+	// source file
+	source string
+	// destination file
+	destination string
+
+	sourceReader io.Reader
+	destWriter   io.Writer
+}
+
+var convert3Example = templates.Examples(`
+	# Convert an Ignition Spec 2 pointer configuration to Spec 3
+	oc ignition convert3 < worker.ign > worker.ign3
+`)
+
+// NewCommandConvert3 creates the command
+func NewCommandConvert3(streams genericclioptions.IOStreams) *cobra.Command {
+	o := convert3Options{
+		sourceReader: streams.In,
+		destWriter:   streams.Out,
+	}
+	cmd := &cobra.Command{
+		Use:     convert3CommandName,
+		Short:   "Convert Ignition Spec 2 to Spec 3",
+		Example: convert3Example,
+		Args:    cobra.ExactArgs(0),
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(o.validate(args))
+			kcmdutil.CheckErr(o.run())
+		},
+	}
+
+	cmd.Flags().StringVar(&o.source, "in", o.source, "File containing input Ignition to convert. Read from stdin if omitted.")
+	cmd.Flags().StringVar(&o.destination, "out", o.destination, "Path to file for converted Ignition. Written to stdout if omitted.")
+	// autocompletion hints
+	cmd.MarkFlagFilename("in")
+	cmd.MarkFlagFilename("out")
+
+	return cmd
+}
+
+func (o *convert3Options) validate(args []string) error {
+	if len(o.source) == 0 && o.sourceReader == nil {
+		return errors.New("an input file, or reader is required")
+	}
+
+	if len(o.destination) == 0 && o.destWriter == nil {
+		return errors.New("an output file or writer is required")
+	}
+	return nil
+}
+
+func (o *convert3Options) run() error {
+	var data []byte
+	switch {
+	case len(o.source) > 0:
+		d, err := ioutil.ReadFile(o.source)
+		if err != nil {
+			return err
+		}
+		data = d
+	case o.sourceReader != nil:
+		d, err := ioutil.ReadAll(o.sourceReader)
+		if err != nil {
+			return err
+		}
+		data = d
+	}
+
+	cfg, rpt, err := ignv2.Parse(data)
+	fmt.Fprintf(os.Stderr, "%s", rpt.String())
+	if err != nil || rpt.IsFatal() {
+		return errors.Errorf("Error parsing spec v2 config: %v\n%v", err, rpt)
```

Comment:
If a user feeds a v3 config to the command, it returns this same error.

```
$ jq .ignition.version < ignitionv3.json 
"3.1.0"
$ ./oc ignition convert3 --in ignitionv3.json --out ignitionv3a.json
error: Error parsing spec v2 config: unsupported config version
```

Could we check the incoming config version before trying to parse?

Reply:
> Could we check the incoming config version before trying to parse?

Not really, to get the version we need to parse...  I think you're just asking for a better error message?  Makes total sense but I think the fix would probably be in ign-converter.

Reply:
Ignition now has `config.util.GetConfigVersion()`, which should work with 2.x configs also.